[refactor]: introduce `Torii`, `Signer`, lightweight `Client` and many tools (#124)
hyperledger/iroha-javascript: packages/client/CHANGELOG.md

# @iroha2/client

## 3.0.0

### Major Changes

- 3b0db98: **refactor!**: split the library functionality into `Torii`, `Signer` and `Client` entities.

>   **What the breaking change is**: previously, `Client` class did everything - constructed transactions payloads, signed them, submitted them and queries, etc. Now some functionality moved out from it.
> 
>   `Torii` does everything related to HTTP/WebSocket communication with Iroha Peer. `Signer` makes signatures. `Client` just wraps them together for tiny convenient methods. Utilities, related to making payloads, signing them, wrapping into final containers etc are exported from the library as separate functions.
> 
>   **Why the change was made**: there are request to extend library functionality, and this change was made to make codebase more scalable. Also, separating crypto functionality from transport stuff is a common practice for blockchain SDKs, e.g. take a look at [`ethers` project](https://docs.ethers.io/v5/).
> 
>   **How a consumer should update their code**: look at `Torii`, `Signer` and `Client` type definitions. They are not very complex and you should adopt very quickly.

Reviewer: `just wraps them together` -- what `them` refers to? signatures?

```suggestion
  ##### What is the change
  
  Previously, the `Client` class did everything: constructed transaction payloads, signed and submitted them, submitted queries, and so on. Now some of this functionality is handled by other classes.

  `Torii` does everything related to HTTP/WebSocket communication with Iroha Peer. `Signer` makes signatures. `Client` just wraps them together for convenient methods. Utilities (e.g. making payloads, signing them, wrapping into final containers) are exported from the library as separate functions.

  ##### Why the change
  
  This change is based on a request to extend library functionality and make codebase more scalable. Separating crypto and transport functionality is a common practice for blockchain SDKs (e.g. take a look at [`ethers` project](https://docs.ethers.io/v5/)).

  ##### How to update your code
  
  Look at `Torii`, `Signer` and `Client` type definitions. They are not very complex and you should adopt very quickly.
```

About the last line -- I think you need to provide something helpful, e.g. an example of the changes that the users would need to make.

Author: > just wraps them up

I mean, `Client` is a wrapper for `Torii` & `Signer`

Reviewer: ok, I'll edit my suggestion them

Author: I am not sure using markdown headers here is an appropriate way, because they are not a subheaders of the major change list item, but of `### Major changes` in general, which breaks semantics IMO.

Reviewer: I checked one of the previous prs with similar changes and made the same suggestion here. You can leave the bold markup but I'd still suggest having new lines around these almost-headers.

Author: @outoftardis,

> About the last line -- I think you need to provide something helpful, e.g. an example of the changes that the users would need to make.

Done.

# @iroha2/client

## 3.0.0

### Major Changes

- 3b0db98: **refactor!**: split the library functionality into `Torii`, `Signer` and `Client` entities.

  ##### What is the change

  Previously, the `Client` class did everything: constructed transaction payloads, signed and submitted them, submitted queries, and so on. Now some of this functionality is handled by other classes.

  `Torii` does everything related to HTTP/WebSocket communication with Iroha Peer. `Signer` makes signatures. `Client` only wraps `Torii` and `Signer` and combines them together to create convenient methods. Utilities (e.g. making payloads, signing them, wrapping into final containers) are exported from the library as separate functions.

  ##### Why the change

  This change is based on a request to extend library functionality and make codebase more scalable. Separating crypto and transport functionality is a common practice for blockchain SDKs (e.g. take a look at [`ethers` project](https://docs.ethers.io/v5/)).

  ##### How to update your code

  Look at `Torii`, `Signer` and `Client` type definitions. They are not very complex and you should adopt very quickly.

>   Example changes:

Reviewer: ```suggestion
```

  ```ts
  import { Client } from '@iroha2/client'

  const client = new Client({
    torii: {
      apiURL: 'http://127.0.0.1:8080',
      telemetryURL: 'http://127.0.0.1:8081',
    },
    accountId,
    keyPair,
    fetch,
    ws,
  })

  // replace with

  import { Client, Torii, Signer } from '@iroha2/client'

  const signer = new Signer(accountId, keyPair)

  const torii = new Torii({
    apiURL: 'http://127.0.0.1:8080',
    telemetryURL: 'http://127.0.0.1:8081',
    fetch,
    ws,
  })

  const client = new Client({ torii, signer })

>   ```

Author: Applied.

  ```ts
  await client.submit(executable)
  await client.request(queryBox)
  await client.getHealth()
  await client.listenForEvents({ filter })

  // replace with

  await client.submitExecutable(executable)
  await client.requestWithQueryBox(queryBox)
  await client.torii.getHealth()
  await client.torii.listenForEvents({ filter })

>   ```

Reviewer: similarly to my suggestion above, spit the code and add text that tells what is happening in these code blocks
